lib/libc/gen/confstr.3
.\" Copyright (c) 1993 The Regents of the University of California.
.\" All rights reserved.
.\"
.\" Redistribution and use in source and binary forms, with or without
.\" modification, are permitted provided that the following conditions
.\" are met:
.\" 1. Redistributions of source code must retain the above copyright
.\"    notice, this list of conditions and the following disclaimer.
.\" 2. Redistributions in binary form must reproduce the above copyright
.\"    notice, this list of conditions and the following disclaimer in the
.\"    documentation and/or other materials provided with the distribution.
.\" 3. All advertising materials mentioning features or use of this software
.\"    must display the following acknowledgement:
.\"	This product includes software developed by the University of
.\"	California, Berkeley and its contributors.
.\" 4. Neither the name of the University nor the names of its contributors
.\"    may be used to endorse or promote products derived from this software
.\"    without specific prior written permission.
.\"
.\" THIS SOFTWARE IS PROVIDED BY THE REGENTS AND CONTRIBUTORS ``AS IS'' AND
.\" ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE
.\" IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE
.\" ARE DISCLAIMED.  IN NO EVENT SHALL THE REGENTS OR CONTRIBUTORS BE LIABLE
.\" FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL
.\" DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS
.\" OR SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION)
.\" HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT
.\" LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY
.\" OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF
.\" SUCH DAMAGE.
.\"
.\"	@(#)confstr.3	5.1 (Berkeley) 05/24/93
.\"
.Dd 
.Dt CONFSTR 3
.Os BSD 4
.Sh NAME
.Nm confstr
.Nd get string-valued configurable variables
.Sh SYNOPSIS
.Fd #include <unistd.h>
.Ft size_t
.Fn confstr "int name" "char *buf" "size_t len"
.Sh DESCRIPTION
.Bf -symbolic
This interface is obsoleted by sysctl(3).
.Ef
.Pp
The
.Fn confstr
function provides a method for applications to get configuration
defined string values.
.Pp
The
.Fa name
argument specifies the system variable to be queried.
Symbolic constants for each name value are found in the include file
.Li <unistd.h> .
The
.Fa len
argument specifies the size of the buffer referenced by the
argument
.Fa buf .
If
.Fa len
is non-zero,
.Fa buf
is a non-null pointer, and
.Fa name
has a value, up to
.Fa len
\- 1 bytes of the value are copied into the buffer 
.Fa buf .
The copied value is always null terminated.
.Pp
The available values are as follows:
.Pp
.Bl -tag -width "123456"
.Pp
.It Li _CS_PATH
Return a value for the
.Ev PATH
environment variable that finds all the standard utilities.
.El
.Sh RETURN VALUES
If the call to
.Nm confstr
is not successful, \-1 is returned and
.Va errno
is set appropriately.
Otherwise, if the variable does not have a configuration defined value,
0 is returned and
.Va errno
is not modified.
Otherwise, the buffer size needed to hold the entire configuration-defined
value is returned.
If this size is greater than the argument
.Fa len ,
the string in
.Fa buf
was truncated.
.Sh ERRORS
The
.Nm confstr
function may fail and set
.Va error
for any of the errors specified for the library functions
.Xr malloc 3
and
.Xr sysctl 3 .
.Pp
In addition, the following errors may be reported:
.Bl -tag -width Er
.It Bq Er EINVAL
The value of the
.Fa name
argument is invalid.
.Sh SEE ALSO
.Xr sysctl 3
.Sh HISTORY
The
.Nm confstr
function first appeared in 4.4BSD.


lib/libc/gen/confstr.c
/*-
 * Copyright (c) 1993 The Regents of the University of California.
 * All rights reserved.
 *
 * Redistribution and use in source and binary forms, with or without
 * modification, are permitted provided that the following conditions
 * are met:
 * 1. Redistributions of source code must retain the above copyright
 *    notice, this list of conditions and the following disclaimer.
 * 2. Redistributions in binary form must reproduce the above copyright
 *    notice, this list of conditions and the following disclaimer in the
 *    documentation and/or other materials provided with the distribution.
 * 3. All advertising materials mentioning features or use of this software
 *    must display the following acknowledgement:
 *	This product includes software developed by the University of
 *	California, Berkeley and its contributors.
 * 4. Neither the name of the University nor the names of its contributors
 *    may be used to endorse or promote products derived from this software
 *    without specific prior written permission.
 *
 * THIS SOFTWARE IS PROVIDED BY THE REGENTS AND CONTRIBUTORS ``AS IS'' AND
 * ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE
 * IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE
 * ARE DISCLAIMED.  IN NO EVENT SHALL THE REGENTS OR CONTRIBUTORS BE LIABLE
 * FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL
 * DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS
 * OR SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION)
 * HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT
 * LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY
 * OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF
 * SUCH DAMAGE.
 */

#if defined(LIBC_SCCS) && !defined(lint)
static char sccsid[] = "@(#)confstr.c	5.1 (Berkeley) 05/03/93";
#endif /* LIBC_SCCS and not lint */

#include <errno.h>
#include <paths.h>
#include <unistd.h>

size_t
confstr(name, buf, len)
	int name;
	char *buf;
	size_t len;
{
	switch (name) {
	case _CS_PATH:
		if (len != 0 && buf != NULL) {
			(void)strncpy(buf, _PATH_DEFPATH, len - 1);
			buf[len - 1] = '\0';
		}
		return (sizeof(_PATH_DEFPATH));
	default:
		errno = EINVAL;
		return (0);
	}
	/* NOTREACHED */
}
